File: client/src/api/players.ts
import { apiClient } from "@/api/client";

export interface PlayerSearchResult {
  id: number;
  name: string;
  score: number;
  image_url: string;
  years_active: string;
}

export interface SavedPlayer {
  id: number;
  name: string;
  image_url?: string;
  years_active?: string;
  [key: string]: any; // Allow additional fields
}

export interface SeasonStats {
  // Basic counting stats
  games: number;
  plate_appearances: number;
  at_bats: number;
  hits: number;
  singles: number;
  doubles: number;
  triples: number;
  home_runs: number;
  runs: number;
  rbi: number;
  walks: number;
  strikeouts: number;
  stolen_bases: number;
  caught_stealing: number;
  
  // Rate stats
  batting_average: number;
  on_base_percentage: number;
  slugging_percentage: number;
  ops: number;
  isolated_power: number;
  babip: number;
  
  // Plate discipline
  walk_rate: number;
  strikeout_rate: number;
  bb_k_ratio: number;
  
  // Advanced metrics
  woba: number;
  wrc_plus: number;
  war: number;
  off: number;
  def_: number;
  base_running: number;
  
  // Contact quality (may be null for older seasons)
  hard_hit_rate?: number | null;
  barrel_rate?: number | null;
  avg_exit_velocity?: number | null;
  avg_launch_angle?: number | null;
  
  // Team
  team_abbrev?: string | null;
}

export interface PlayerDetail {
  mlbam_id: number;
  fangraphs_id: number;
  name: string;
  image_url: string;
  years_active: string;
  team_abbrev?: string | null;
  overall_score: number;
  seasons: Record<string, SeasonStats>;  // Year -> Stats mapping
}

export interface AddPlayerResponse {
  message: string;
  player_id: string;
}

export interface DeletePlayerResponse {
  message: string;
}

export const playersApi = {
  search: async (query: string): Promise<PlayerSearchResult[]> => {
    return apiClient.get<PlayerSearchResult[]>(`/api/players/search?q=${encodeURIComponent(query)}`);
  },

  getDetail: async (playerId: number): Promise<PlayerDetail> => {
    return apiClient.get<PlayerDetail>(`/api/players/${playerId}/detail`);
  },

  getSaved: async (): Promise<SavedPlayer[]> => {
    return apiClient.get<SavedPlayer[]>("/api/players/saved");
  },

  addSaved: async (player: { id: number; name: string; image_url?: string; years_active?: string }): Promise<AddPlayerResponse> => {
    return apiClient.post<AddPlayerResponse>("/api/players/saved", player);
  },

  deleteSaved: async (playerId: number): Promise<DeletePlayerResponse> => {
    return apiClient.delete<DeletePlayerResponse>(`/api/players/saved/${playerId}`);
  },
};

File: server/models/players.py
from pydantic import BaseModel
from typing import Optional, Dict

class PlayerSearchResult(BaseModel):
    """Player search result from the index"""
    id: int
    name: str
    score: float
    image_url: str
    years_active: str

class AddPlayerResponse(BaseModel):
    """Response after adding a player"""
    message: str
    player_id: str

class DeletePlayerResponse(BaseModel):
    """Response after deleting a player"""
    message: str

class SavedPlayer(BaseModel):
    """Saved player data"""
    id: int
    name: str
    image_url: Optional[str] = None
    years_active: Optional[str] = None
    
    class Config:
        extra = "allow"  # Allow additional fields from Firestore

class SeasonStats(BaseModel):
    """Stats for a single season"""
    # Basic counting stats
    games: int = 0
    plate_appearances: int = 0
    at_bats: int = 0
    hits: int = 0
    singles: int = 0
    doubles: int = 0
    triples: int = 0
    home_runs: int = 0
    runs: int = 0
    rbi: int = 0
    walks: int = 0
    strikeouts: int = 0
    stolen_bases: int = 0
    caught_stealing: int = 0
    
    # Rate stats
    batting_average: float = 0.0
    on_base_percentage: float = 0.0
    slugging_percentage: float = 0.0
    ops: float = 0.0
    isolated_power: float = 0.0
    babip: float = 0.0
    
    # Plate discipline
    walk_rate: float = 0.0
    strikeout_rate: float = 0.0
    bb_k_ratio: float = 0.0
    
    # Advanced metrics
    woba: float = 0.0
    wrc_plus: float = 0.0
    war: float = 0.0
    off: float = 0.0
    def_: float = 0.0  # 'def' is a Python keyword, use def_
    base_running: float = 0.0
    
    # Contact quality (may be None for older seasons)
    hard_hit_rate: Optional[float] = None
    barrel_rate: Optional[float] = None
    avg_exit_velocity: Optional[float] = None
    avg_launch_angle: Optional[float] = None
    
    # Team
    team_abbrev: Optional[str] = None
    
    class Config:
        extra = "allow"

class PlayerDetail(BaseModel):
    """Detailed player information with all seasons stats"""
    mlbam_id: int
    fangraphs_id: int
    name: str
    image_url: str
    years_active: str
    team_abbrev: Optional[str] = None
    overall_score: float = 0.0
    seasons: Dict[str, SeasonStats]  # Year -> Stats mapping
    
    class Config:
        extra = "allow"

